File: proof/csharp/src/MathUtils.cs
// Simple math utility class for Regrets C# capture/validate demonstration.
namespace MathLib;

public static class MathUtils
{
    /// <summary>
    /// Add two integers and return the sum.
    /// </summary>
    public static int Add(int a, int b) => a + b;

    /// <summary>
    /// Multiply two integers and return the product.
    /// </summary>
    public static int Multiply(int a, int b) => a * b;

    /// <summary>
    /// Reverse a string.
    /// </summary>
    public static string ReverseString(string input)
    {
        if (string.IsNullOrEmpty(input)) return input ?? "";
        var chars = input.ToCharArray();
        Array.Reverse(chars);
        return new string(chars);
    }

    /// <summary>
    /// Check if a number is prime.
    /// </summary>
    public static bool IsPrime(int n)
    {
        if (n < 2) return false;
        if (n == 2) return true;
        if (n % 2 == 0) return false;
        for (int i = 3; i * i <= n; i += 2)
        {
            if (n % i == 0) return false;
        }
        return true;
    }
}

File: proof/dart/math_utils.dart
/// math_utils.dart — example pure functions for Regrets fingerprint testing

/// Add two integers
int add(int a, int b) => a + b;

/// Multiply two integers
int mul(int a, int b) => a * b;

/// Compute factorial recursively
int factorial(int n) {
  if (n <= 1) return 1;
  return n * factorial(n - 1);
}

/// Check if a number is prime
bool isPrime(int n) {
  if (n < 2) return false;
  if (n < 4) return true;
  if (n % 2 == 0 || n % 3 == 0) return false;
  int i = 5;
  while (i * i <= n) {
    if (n % i == 0 || n % (i + 2) == 0) return false;
    i += 6;
  }
  return true;
}

/// Reverse a string
String reverseString(String s) => s.split('').reversed.join('');

/// Fibonacci number at position n (0-indexed)
int fibonacci(int n) {
  if (n <= 0) return 0;
  if (n == 1) return 1;
  int a = 0, b = 1;
  for (int i = 2; i <= n; i++) {
    final temp = a + b;
    a = b;
    b = temp;
  }
  return b;
}
